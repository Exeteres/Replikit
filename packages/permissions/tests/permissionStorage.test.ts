import {
    PermissionStorage,
    RoleNotFoundError,
    InvalidFallbackRoleError
} from "@replikit/permissions";

describe("PermissionStorage", () => {
    it("should add a list of permissions", () => {
        const storage = new PermissionStorage();
        storage.addPermissions("user", ["test1", "test2"]);

        expect(storage).toMatchSnapshot();
    });

    it("should add a role with permissions", () => {
        const storage = new PermissionStorage();
        storage.updateRole("user", "Admin", {
            permissions: ["test1", "test2"]
        });

        expect(storage).toMatchSnapshot();
    });

    it("should add a role with fallbackRole", () => {
        const storage = new PermissionStorage();
        storage.updateRole("user", "Moderator");
        storage.updateRole("user", "Admin", {
            fallbackRoles: ["Moderator"]
        });

        expect(storage).toMatchSnapshot();
    });

    it("should throw an error if fallbackRole not exists", () => {
        const storage = new PermissionStorage();
        expect(() => {
            storage.updateRole("user", "Admin", {
                fallbackRoles: ["Moderator"]
            });
        }).toThrow(RoleNotFoundError);
    });

    it("should throw an error if fallbackRole matches the primary role name", () => {
        const storage = new PermissionStorage();
        expect(() => {
            storage.updateRole("user", "Admin", {
                fallbackRoles: ["Admin"]
            });
        }).toThrow(InvalidFallbackRoleError);
    });

    it("should update the existing role", () => {
        const storage = new PermissionStorage();
        storage.updateRole("user", "Moderator");
        storage.updateRole("user", "Admin", {
            permissions: ["test1", "test2"]
        });
        storage.updateRole("user", "Admin", {
            permissions: ["test3"],
            fallbackRoles: ["Moderator"]
        });

        expect(storage).toMatchSnapshot();
    });
});
