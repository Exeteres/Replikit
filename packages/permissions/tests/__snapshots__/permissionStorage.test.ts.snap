// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionStorage should add a list of permissions 1`] = `
PermissionStorage {
  "permissionMap": Map {
    "user" => Array [
      "test1",
      "test2",
    ],
  },
  "roles": Array [],
}
`;

exports[`PermissionStorage should add a role with fallbackRole 1`] = `
PermissionStorage {
  "permissionMap": Map {},
  "roles": Array [
    Object {
      "fallbackRoles": Array [],
      "name": "Moderator",
      "permissions": Array [],
      "type": "user",
    },
    Object {
      "fallbackRoles": Array [
        Object {
          "fallbackRoles": Array [],
          "name": "Moderator",
          "permissions": Array [],
          "type": "user",
        },
      ],
      "name": "Admin",
      "permissions": Array [],
      "type": "user",
    },
  ],
}
`;

exports[`PermissionStorage should add a role with permissions 1`] = `
PermissionStorage {
  "permissionMap": Map {},
  "roles": Array [
    Object {
      "fallbackRoles": Array [],
      "name": "Admin",
      "permissions": Array [
        "test1",
        "test2",
      ],
      "type": "user",
    },
  ],
}
`;

exports[`PermissionStorage should update the existing role 1`] = `
PermissionStorage {
  "permissionMap": Map {},
  "roles": Array [
    Object {
      "fallbackRoles": Array [],
      "name": "Moderator",
      "permissions": Array [],
      "type": "user",
    },
    Object {
      "fallbackRoles": Array [
        Object {
          "fallbackRoles": Array [],
          "name": "Moderator",
          "permissions": Array [],
          "type": "user",
        },
      ],
      "name": "Admin",
      "permissions": Array [
        "test1",
        "test2",
        "test3",
      ],
      "type": "user",
    },
  ],
}
`;
